File: backend/app/services/token_optimization_service.py
"""
Token Optimization Service
Optimizes Firebase token handling for socket connections to prevent parse errors.
"""

import time
import json
from typing import Dict, Any, Optional, Tuple, List
from app.utils.railway_logger import railway_logger
from app.utils.firebase_token_analyzer import firebase_token_analyzer

class TokenOptimizationService:
    """Service for optimizing Firebase token handling in socket connections."""
    
    def __init__(self):
        self.token_cache = {}  # Cache for validated tokens
        self.token_validation_history = {}  # Track validation attempts
        self.optimization_stats = {
            'tokens_validated': 0,
            'tokens_cached': 0,
            'cache_hits': 0,
            'validation_failures': 0,
            'optimization_applied': 0
        }
        self.start_time = time.time()
    
    def validate_token_for_socket(self, token: str, user_id: str = 'unknown') -> Dict[str, Any]:
        """Validate a token for socket usage with optimization."""
        try:
            # Check cache first
            cache_key = f"{user_id}:{hash(token)}"
            if cache_key in self.token_cache:
                cached_result = self.token_cache[cache_key]
                if time.time() - cached_result['timestamp'] < 300:  # 5 minute cache
                    self.optimization_stats['cache_hits'] += 1
                    railway_logger.log('token_optimization', 10, f"Token cache hit for user: {user_id}")
                    return cached_result['result']
            
            # Analyze token for potential issues
            analysis = firebase_token_analyzer.analyze_token(token, f"socket_validation_{user_id}")
            
            # Create validation result
            validation_result = {
                'is_valid': analysis.get('is_valid_format', False),
                'has_issues': analysis.get('has_parse_issues', False),
                'issues': analysis.get('issues', []),
                'token_size': analysis.get('token_size_bytes', 0),
                'analysis': analysis,
                'optimization_applied': False,
                'recommendations': []
            }
            
            # Apply optimizations if needed
            if validation_result['has_issues']:
                optimization_result = self._apply_token_optimizations(token, analysis)
                validation_result['optimization_applied'] = optimization_result['applied']
                validation_result['recommendations'] = optimization_result['recommendations']
                
                if optimization_result['applied']:
                    self.optimization_stats['optimization_applied'] += 1
            
            # Cache the result
            self.token_cache[cache_key] = {
                'result': validation_result,
                'timestamp': time.time()
            }
            self.optimization_stats['tokens_cached'] += 1
            
            # Update statistics
            self.optimization_stats['tokens_validated'] += 1
            if not validation_result['is_valid']:
                self.optimization_stats['validation_failures'] += 1
            
            # Track validation history
            self.token_validation_history[user_id] = {
                'last_validation': time.time(),
                'validation_count': self.token_validation_history.get(user_id, {}).get('validation_count', 0) + 1,
                'last_result': validation_result
            }
            
            railway_logger.log('token_optimization', 10, f"Token validated for user: {user_id}, valid: {validation_result['is_valid']}")
            
            return validation_result
            
        except Exception as e:
            railway_logger.log('token_optimization', 40, f"Token validation failed for user {user_id}: {str(e)}")
            return {
                'is_valid': False,
                'has_issues': True,
                'issues': [f"Validation error: {str(e)}"],
                'token_size': 0,
                'analysis': None,
                'optimization_applied': False,
                'recommendations': ['Fix token validation error']
            }
    
    def _apply_token_optimizations(self, token: str, analysis: Dict[str, Any]) -> Dict[str, Any]:
        """Apply optimizations to a token if possible."""
        optimizations = {
            'applied': False,
            'recommendations': []
        }
        
        try:
            # Check for common issues and provide recommendations
            issues = analysis.get('issues', [])
            
            for issue in issues:
                if 'Token too short' in issue:
                    optimizations['recommendations'].append('Token appears to be invalid or corrupted')
                elif 'Token unusually long' in issue:
                    optimizations['recommendations'].append('Consider token refresh to get a shorter token')
                elif 'control characters' in issue.lower():
                    optimizations['recommendations'].append('Token contains invalid characters, refresh required')
                elif 'non-ASCII' in issue.lower():
                    optimizations['recommendations'].append('Token contains non-ASCII characters, refresh required')
                elif 'serialization failed' in issue.lower():
                    optimizations['recommendations'].append('Token cannot be serialized, refresh required')
            
            # Check token expiry
            token_parts = analysis.get('token_parts', {})
            payload = token_parts.get('payload_decoded', {})
            if payload:
                exp_time = payload.get('exp')
                if exp_time:
                    current_time = time.time()
                    if exp_time < current_time:
                        optimizations['recommendations'].append('Token has expired, refresh required')
                    elif exp_time - current_time < 300:  # Less than 5 minutes
                        optimizations['recommendations'].append('Token expires soon, consider refresh')
            
            # If we have specific recommendations, mark as optimization applied
            if optimizations['recommendations']:
                optimizations['applied'] = True
            
        except Exception as e:
            railway_logger.log('token_optimization', 30, f"Token optimization failed: {str(e)}")
            optimizations['recommendations'].append(f"Optimization error: {str(e)}")
        
        return optimizations
    
    def optimize_socket_message_with_token(self, message: Dict[str, Any], token: str, user_id: str = 'unknown') -> Dict[str, Any]:
        """Optimize a socket message containing a token."""
        try:
            # Validate token first
            token_validation = self.validate_token_for_socket(token, user_id)
            
            # Create optimized message
            optimized_message = message.copy()
            
            # Apply token-specific optimizations
            if token_validation['has_issues']:
                # If token has issues, we might want to handle it differently
                optimized_message['_token_issues'] = token_validation['issues']
                optimized_message['_token_optimization_applied'] = True
            
            # Ensure token is properly formatted in message
            if 'id_token' in optimized_message:
                optimized_message['id_token'] = token
            
            # Add token metadata for debugging
            optimized_message['_token_metadata'] = {
                'size': token_validation['token_size'],
                'validated_at': time.time(),
                'has_issues': token_validation['has_issues']
            }
            
            railway_logger.log('token_optimization', 10, f"Socket message optimized for user: {user_id}")
            
            return optimized_message
            
        except Exception as e:
            railway_logger.log('token_optimization', 40, f"Socket message optimization failed for user {user_id}: {str(e)}")
            return message  # Return original message if optimization fails
    
    def handle_token_refresh(self, old_token: str, new_token: str, user_id: str = 'unknown') -> Dict[str, Any]:
        """Handle token refresh and analyze the impact."""
        try:
            # Analyze both tokens
            refresh_analysis = firebase_token_analyzer.analyze_token_refresh_timing(old_token, new_token)
            
            # Validate new token
            new_token_validation = self.validate_token_for_socket(new_token, user_id)
            
            # Clear old token from cache
            old_cache_key = f"{user_id}:{hash(old_token)}"
            if old_cache_key in self.token_cache:
                del self.token_cache[old_cache_key]
            
            # Create refresh result
            refresh_result = {
                'refresh_successful': new_token_validation['is_valid'],
                'old_token_analysis': refresh_analysis['old_token_analysis'],
                'new_token_analysis': refresh_analysis['new_token_analysis'],
                'refresh_impact': refresh_analysis['refresh_impact'],
                'recommendations': refresh_analysis['recommendations'] + new_token_validation['recommendations'],
                'optimization_applied': new_token_validation['optimization_applied']
            }
            
            railway_logger.log('token_optimization', 10, f"Token refresh handled for user: {user_id}")
            
            return refresh_result
            
        except Exception as e:
            railway_logger.log('token_optimization', 40, f"Token refresh handling failed for user {user_id}: {str(e)}")
            return {
                'refresh_successful': False,
                'error': str(e),
                'recommendations': ['Fix token refresh error']
            }
    
    def get_optimization_statistics(self) -> Dict[str, Any]:
        """Get token optimization statistics."""
        cache_size = len(self.token_cache)
        validation_history_size = len(self.token_validation_history)
        
        # Calculate cache hit rate
        total_cache_requests = self.optimization_stats['tokens_validated']
        cache_hit_rate = (self.optimization_stats['cache_hits'] / total_cache_requests) if total_cache_requests > 0 else 0
        
        # Calculate validation success rate
        validation_success_rate = 1 - (self.optimization_stats['validation_failures'] / total_cache_requests) if total_cache_requests > 0 else 1
        
        return {
            'optimization_stats': self.optimization_stats.copy(),
            'cache_size': cache_size,
            'validation_history_size': validation_history_size,
            'cache_hit_rate': cache_hit_rate,
            'validation_success_rate': validation_success_rate,
            'optimization_rate': (self.optimization_stats['optimization_applied'] / total_cache_requests) if total_cache_requests > 0 else 0,
            'uptime_seconds': time.time() - self.start_time
        }
    
    def clear_token_cache(self, user_id: Optional[str] = None) -> None:
        """Clear token cache for a specific user or all users."""
        if user_id:
            # Clear cache entries for specific user
            keys_to_remove = [key for key in self.token_cache.keys() if key.startswith(f"{user_id}:")]
            for key in keys_to_remove:
                del self.token_cache[key]
            railway_logger.log('token_optimization', 10, f"Token cache cleared for user: {user_id}")
        else:
            # Clear entire cache
            self.token_cache.clear()
            railway_logger.log('token_optimization', 10, "Token cache cleared for all users")
    
    def get_token_recommendations(self, user_id: str) -> List[str]:
        """Get token recommendations for a specific user."""
        recommendations = []
        
        # Check user's validation history
        user_history = self.token_validation_history.get(user_id, {})
        if user_history:
            last_result = user_history.get('last_result', {})
            if last_result.get('has_issues', False):
                recommendations.extend(last_result.get('recommendations', []))
            
            validation_count = user_history.get('validation_count', 0)
            if validation_count > 10:  # User has many validation attempts
                recommendations.append('Consider implementing token refresh optimization')
        
        # Get general recommendations from token analyzer
        analyzer_recommendations = firebase_token_analyzer.get_recommendations()
        recommendations.extend(analyzer_recommendations)
        
        return recommendations
    
    def reset_statistics(self) -> None:
        """Reset optimization statistics."""
        self.optimization_stats = {
            'tokens_validated': 0,
            'tokens_cached': 0,
            'cache_hits': 0,
            'validation_failures': 0,
            'optimization_applied': 0
        }
        self.token_validation_history.clear()
        self.start_time = time.time()
        railway_logger.log('token_optimization', 10, "Token optimization statistics reset")

# Global instance
token_optimization_service = TokenOptimizationService()

File: frontend/src/services/tokenOptimizationService.ts
/**
 * Token Optimization Service
 * Handles Firebase token validation, optimization, and refresh for socket connections.
 */

import { authService } from './authService'
import { canvasAPI } from './api'

export interface TokenValidationResult {
  isValid: boolean
  hasIssues: boolean
  issues: string[]
  tokenSize: number
  recommendations: string[]
  optimizationApplied: boolean
}

export interface TokenRefreshResult {
  refreshSuccessful: boolean
  oldTokenAnalysis: any
  newTokenAnalysis: any
  refreshImpact: any
  recommendations: string[]
  optimizationApplied: boolean
}

class TokenOptimizationService {
  private tokenCache = new Map<string, { result: TokenValidationResult; timestamp: number }>()
  private refreshInProgress = false
  private validationStats = {
    tokensValidated: 0,
    cacheHits: 0,
    validationFailures: 0,
    optimizationApplied: 0
  }

  /**
   * Validate a token for socket usage
   */
  async validateTokenForSocket(token: string, userId: string = 'unknown'): Promise<TokenValidationResult> {
    try {
      // Check cache first (5 minute cache)
      const cacheKey = `${userId}:${this.hashToken(token)}`
      const cached = this.tokenCache.get(cacheKey)
      if (cached && Date.now() - cached.timestamp < 300000) {
        this.validationStats.cacheHits++
        console.log('Token cache hit for user:', userId)
        return cached.result
      }

      // Validate token locally first
      const localValidation = this.validateTokenLocally(token)
      
      // If local validation passes, validate with backend
      if (localValidation.isValid) {
        try {
          const backendValidation = await this.validateTokenWithBackend(token, userId)
          if (backendValidation.isValid) {
            // Cache the result
            this.tokenCache.set(cacheKey, {
              result: backendValidation,
              timestamp: Date.now()
            })
            
            this.validationStats.tokensValidated++
            return backendValidation
          } else {
            this.validationStats.validationFailures++
            return backendValidation
          }
        } catch (error) {
          console.warn('Backend token validation failed, using local validation:', error)
          this.validationStats.validationFailures++
          return localValidation
        }
      } else {
        this.validationStats.validationFailures++
        return localValidation
      }
    } catch (error) {
      console.error('Token validation failed:', error)
      this.validationStats.validationFailures++
      return {
        isValid: false,
        hasIssues: true,
        issues: [`Validation error: ${error}`],
        tokenSize: 0,
        recommendations: ['Fix token validation error'],
        optimizationApplied: false
      }
    }
  }

  /**
   * Validate token locally without backend call
   */
  private validateTokenLocally(token: string): TokenValidationResult {
    const issues: string[] = []
    let isValid = true

    // Basic token format validation
    if (!token || typeof token !== 'string') {
      issues.push('Token is empty or not a string')
      isValid = false
    } else if (token.length < 100) {
      issues.push('Token too short (likely invalid)')
      isValid = false
    } else if (token.length > 10000) {
      issues.push('Token unusually long')
      isValid = false
    }

    // Check for problematic characters
    if (token) {
      for (let i = 0; i < token.length; i++) {
        const char = token[i]
        const charCode = token.charCodeAt(i)
        
        if (charCode < 32 && !['\t', '\n', '\r'].includes(char)) {
          issues.push('Token contains control characters')
          isValid = false
          break
        } else if (charCode > 127) {
          issues.push('Token contains non-ASCII characters')
          isValid = false
          break
        }
      }
    }

    // Test JSON serialization
    try {
      JSON.stringify({ id_token: token })
    } catch (error) {
      issues.push(`Token serialization failed: ${error}`)
      isValid = false
    }

    const tokenSize = token ? new Blob([token]).size : 0
    const recommendations = this.generateRecommendations(issues)

    return {
      isValid,
      hasIssues: issues.length > 0,
      issues,
      tokenSize,
      recommendations,
      optimizationApplied: false
    }
  }

  /**
   * Validate token with backend
   */
  private async validateTokenWithBackend(token: string, userId: string): Promise<TokenValidationResult> {
    try {
      const response = await canvasAPI.post('/api/token-analysis/validate', {
        token,
        user_id: userId
      })

      if (response.data.status === 'success') {
        return response.data.validation
      } else {
        throw new Error(response.data.message || 'Backend validation failed')
      }
    } catch (error) {
      throw new Error(`Backend validation error: ${error}`)
    }
  }

  /**
   * Optimize a socket message containing a token
   */
  async optimizeSocketMessageWithToken(message: Record<string, any>, token: string, userId: string = 'unknown'): Promise<Record<string, any>> {
    try {
      // Validate token first
      const tokenValidation = await this.validateTokenForSocket(token, userId)
      
      // Create optimized message
      const optimizedMessage = { ...message }
      
      // Apply token-specific optimizations
      if (tokenValidation.hasIssues) {
        optimizedMessage._tokenIssues = tokenValidation.issues
        optimizedMessage._tokenOptimizationApplied = true
      }
      
      // Ensure token is properly formatted in message
      if ('id_token' in optimizedMessage) {
        optimizedMessage.id_token = token
      }
      
      // Add token metadata for debugging
      optimizedMessage._tokenMetadata = {
        size: tokenValidation.tokenSize,
        validatedAt: Date.now(),
        hasIssues: tokenValidation.hasIssues
      }
      
      console.log('Socket message optimized for user:', userId)
      return optimizedMessage
      
    } catch (error) {
      console.error('Socket message optimization failed:', error)
      return message // Return original message if optimization fails
    }
  }

  /**
   * Handle token refresh
   */
  async handleTokenRefresh(oldToken: string, newToken: string, userId: string = 'unknown'): Promise<TokenRefreshResult> {
    try {
      if (this.refreshInProgress) {
        throw new Error('Token refresh already in progress')
      }

      this.refreshInProgress = true

      // Clear old token from cache
      const oldCacheKey = `${userId}:${this.hashToken(oldToken)}`
      this.tokenCache.delete(oldCacheKey)

      // Validate new token
      const newTokenValidation = await this.validateTokenForSocket(newToken, userId)

      // Analyze refresh impact
      const oldTokenSize = new Blob([oldToken]).size
      const newTokenSize = newTokenValidation.tokenSize
      const sizeChange = newTokenSize - oldTokenSize
      const sizeChangePercent = oldTokenSize > 0 ? (sizeChange / oldTokenSize) * 100 : 0

      const refreshResult: TokenRefreshResult = {
        refreshSuccessful: newTokenValidation.isValid,
        oldTokenAnalysis: {
          tokenSize: oldTokenSize,
          isValid: true // Assume old token was valid if we're refreshing
        },
        newTokenAnalysis: newTokenValidation,
        refreshImpact: {
          sizeChange,
          sizeChangePercent,
          oldHasIssues: false,
          newHasIssues: newTokenValidation.hasIssues
        },
        recommendations: newTokenValidation.recommendations,
        optimizationApplied: newTokenValidation.optimizationApplied
      }

      // Add refresh-specific recommendations
      if (sizeChangePercent > 50) {
        refreshResult.recommendations.push('Token size increased significantly during refresh')
      }

      if (newTokenValidation.hasIssues) {
        refreshResult.recommendations.push('New token introduced parse issues')
      }

      console.log('Token refresh handled for user:', userId)
      return refreshResult

    } catch (error) {
      console.error('Token refresh handling failed:', error)
      return {
        refreshSuccessful: false,
        oldTokenAnalysis: { error: String(error) },
        newTokenAnalysis: { error: String(error) },
        refreshImpact: { error: String(error) },
        recommendations: ['Fix token refresh error'],
        optimizationApplied: false
      }
    } finally {
      this.refreshInProgress = false
    }
  }

  /**
   * Get current token and validate it
   */
  async getValidatedCurrentToken(): Promise<{ token: string; validation: TokenValidationResult } | null> {
    try {
      const token = await authService.getValidToken()
      if (!token) {
        return null
      }

      const validation = await this.validateTokenForSocket(token)
      return { token, validation }
    } catch (error) {
      console.error('Failed to get validated current token:', error)
      return null
    }
  }

  /**
   * Refresh token if needed
   */
  async refreshTokenIfNeeded(): Promise<{ token: string; refreshed: boolean } | null> {
    try {
      const currentToken = await authService.getValidToken()
      if (!currentToken) {
        return null
      }

      // Check if token needs refresh
      const validation = await this.validateTokenForSocket(currentToken)
      if (validation.isValid && !validation.hasIssues) {
        return { token: currentToken, refreshed: false }
      }

      // Try to refresh token
      const newToken = await authService.publicRefreshToken()
      if (newToken && newToken !== currentToken) {
        const refreshResult = await this.handleTokenRefresh(currentToken, newToken)
        return { 
          token: newToken, 
          refreshed: refreshResult.refreshSuccessful 
        }
      }

      return { token: currentToken, refreshed: false }
    } catch (error) {
      console.error('Token refresh failed:', error)
      return null
    }
  }

  /**
   * Generate recommendations based on issues
   */
  private generateRecommendations(issues: string[]): string[] {
    const recommendations: string[] = []

    for (const issue of issues) {
      if (issue.includes('too short')) {
        recommendations.push('Token appears to be invalid or corrupted')
      } else if (issue.includes('unusually long')) {
        recommendations.push('Consider token refresh to get a shorter token')
      } else if (issue.includes('control characters')) {
        recommendations.push('Token contains invalid characters, refresh required')
      } else if (issue.includes('non-ASCII')) {
        recommendations.push('Token contains non-ASCII characters, refresh required')
      } else if (issue.includes('serialization failed')) {
        recommendations.push('Token cannot be serialized, refresh required')
      }
    }

    return recommendations
  }

  /**
   * Hash token for cache key
   */
  private hashToken(token: string): string {
    let hash = 0
    for (let i = 0; i < token.length; i++) {
      const char = token.charCodeAt(i)
      hash = ((hash << 5) - hash) + char
      hash = hash & hash // Convert to 32-bit integer
    }
    return hash.toString()
  }

  /**
   * Clear token cache
   */
  clearTokenCache(userId?: string): void {
    if (userId) {
      // Clear cache entries for specific user
      const keysToDelete: string[] = []
      for (const key of this.tokenCache.keys()) {
        if (key.startsWith(`${userId}:`)) {
          keysToDelete.push(key)
        }
      }
      keysToDelete.forEach(key => this.tokenCache.delete(key))
      console.log('Token cache cleared for user:', userId)
    } else {
      // Clear entire cache
      this.tokenCache.clear()
      console.log('Token cache cleared for all users')
    }
  }

  /**
   * Get optimization statistics
   */
  getOptimizationStats() {
    const cacheSize = this.tokenCache.size
    const totalRequests = this.validationStats.tokensValidated
    const cacheHitRate = totalRequests > 0 ? this.validationStats.cacheHits / totalRequests : 0
    const validationSuccessRate = totalRequests > 0 ? 1 - (this.validationStats.validationFailures / totalRequests) : 1

    return {
      ...this.validationStats,
      cacheSize,
      cacheHitRate,
      validationSuccessRate,
      optimizationRate: totalRequests > 0 ? this.validationStats.optimizationApplied / totalRequests : 0
    }
  }

  /**
   * Reset statistics
   */
  resetStats(): void {
    this.validationStats = {
      tokensValidated: 0,
      cacheHits: 0,
      validationFailures: 0,
      optimizationApplied: 0
    }
    this.tokenCache.clear()
    console.log('Token optimization statistics reset')
  }
}

export const tokenOptimizationService = new TokenOptimizationService()
